django/db/models/indexes.py

```diff
@@ -0,0 +1,57 @@
+
+__all__ = ['Index']
+
+
+class Index(object):
+
+    def __init__(self, *fields, **kwargs):
+        self.fields = fields
+        if 'model' in kwargs:
+            self.model = kwargs['model']
+        if 'name' in kwargs:
+            self.name = kwargs['name']
+
+    def as_sql(self, schema, suffix=''):
+        columns = [field for field in self.fields]
+        fields = [self.model._meta.get_field(field) for field in self.fields]
+        if len(fields) == 1 and fields[0].db_tablespace:
+            tablespace_sql = schema.connection.ops.tablespace_sql(fields[0].db_tablespace)
+        elif self.model._meta.db_tablespace:
+            tablespace_sql = schema.connection.ops.tablespace_sql(self.model._meta.db_tablespace)
+        else:
+            tablespace_sql = ""
+        if tablespace_sql:
+            tablespace_sql = " " + tablespace_sql
+
+        if not hasattr(self, 'name'):
+            self.name = schema._create_index_name(self.model, columns, suffix=suffix)
+        quote_name = schema.quote_name
+        return schema.sql_create_index % {
+            "table": quote_name(self.model._meta.db_table),
+            "name": quote_name(self.name),
+            "columns": ", ".join(quote_name(column) for column in columns),
+            "extra": tablespace_sql,
+        }
+
+    def remove_sql(self, schema):
+        quote_name = schema.quote_name
+        return schema.sql_delete_index % {
+            "table": quote_name(self.model._meta.db_table),
+            "name": quote_name(self.name),
```

Comment:
`self.name` is being set in `create_sql` in any case ([here](https://github.com/django/django/pull/6726/files#diff-803276386305dab54c20b97d3d34b82fR63)). So I guess if an index exists in the database, we can safely assume that `self.name` is set. Please let me know your thoughts on this.